readthedocs/oauth/clients.py

```diff
@@ -0,0 +1,71 @@
+from datetime import datetime
+
+import structlog
+from django.utils import timezone
+from requests_oauthlib import OAuth2Session
+
+log = structlog.get_logger(__name__)
+
+
+def _get_token_updater(token):
+    """
+    Update token given data from OAuth response.
+
+    Expect the following response into the closure::
+
+        {
+            u'token_type': u'bearer',
+            u'scopes': u'webhook repository team account',
+            u'refresh_token': u'...',
+            u'access_token': u'...',
+            u'expires_in': 3600,
+            u'expires_at': 1449218652.558185
+        }
+    """
+
+    def _updater(data):
+        token.token = data["access_token"]
+        token.token_secret = data.get("refresh_token", "")
+        token.expires_at = timezone.make_aware(
+            datetime.fromtimestamp(data["expires_at"]),
+        )
+        token.save()
+        log.info("Updated token.", token_id=token.pk)
+
+    return _updater
+
+
+def get_oauth2_client(account):
+    """Get an OAuth2 client for the given social account."""
+    token = account.socialtoken_set.first()
+    if token is None:
+        return None
+
+    token_config = {
+        "access_token": token.token,
+        "token_type": "bearer",
+    }
+    if token.expires_at is not None:
+        token_expires = (token.expires_at - timezone.now()).total_seconds()
+        token_config.update(
+            {
+                "refresh_token": token.token_secret,
+                "expires_in": token_expires,
```

Comment:
I'm a little surprised we're setting this internally -- is it so we can update it if needed?

Reply:
yeah, allauth doesn't provide an oauth2 client, so we use requests_oauthlib, and we need to glue the token from allauth other lib. Most of this code was just moved.